File: zertz/zertz.py
from pyp5js import *
from .marble import Marble
from .ring import Ring
from .settings import *

class Zertz:
    def __init__(self,
                 side_length: int,
                 cell_type: callable):
        self.side_length = side_length
        self.cell_type = cell_type

        self.MARBLE_RADIUS = CANVAS_HEIGHT * 45 / 900
        self.RING_RADIUS = CANVAS_HEIGHT * 65 / 900
        self.marbles = []
        self.rings = []

        self.undo = list()
        self.redo = list()

    def _populate_rings(self):
        for col_id in range(self.num_cells):
            self.rings.append(self.cell_type())
        self.rings.append(None)


    def _get_num_cells(self):
        """
        To calculate total number of grids for a given side length,
        we use the formula
            side_length + (side_length + 1) + ... + 2*side_length -1 + ... (side_length + 1) + side_length
        """
        cells_per_col = list(range(self.side_length, 2*self.side_length))
        self.cells_per_col =  cells_per_col + cells_per_col[::-1][1:]

        # self.cum_cells = []
        # total = 0
        # for i in self.cells_per_col:
        #     self.cum_cells.append(total + i)
        return sum(self.cells_per_col)

    def _get_col_row_from_ind(self, index:int):
        for col, num_cells in enumerate(self.cum_cells):
            if index < num_cells:
                break
        if col == 0:
            row = index
        else:
            row = index - self.cum_cells[col - 1]
        return col, row

    def _get_ind_from_col_row(self, col:int, row:int):
        if col > 0:
            return self.cum_cells[col - 1] + row
        else:
            return row

    def _get_neighbours(self, i:int):
        neighbours = []

        curr_col, curr_row = self._get_col_row_from_ind(i)

        if curr_row > 0:
            neighbours.append(i - 1)
        else:
            neighbours.append(self.num_cells + 1)

        if curr_row < self.cells_per_col[curr_col] - 1:
            neighbours.append(i + 1)
        else:
            neighbours.append(self.num_cells + 1)

        prev_col = curr_col - 1
        next_col = curr_col + 1

        if prev_col > -1 and curr_row < self.cells_per_col[prev_col]:
            neighbours.append(self._get_ind_from_col_row(prev_col, curr_row))
        else:
            neighbours.append(self.num_cells + 1)

        if next_col < len(self.cells_per_col) and curr_row < self.cells_per_col[next_col]:
            neighbours.append(self._get_ind_from_col_row(next_col, curr_row))
        else:
            neighbours.append(self.num_cells + 1)

        if curr_col == len(self.cells_per_col) // 2:
            if prev_col > -1 and curr_row - 1 < self.cells_per_col[prev_col] and curr_row > 0:
                neighbours.append(self._get_ind_from_col_row(prev_col, curr_row - 1))
            else:
                neighbours.append(self.num_cells + 1)

            if next_col < len(self.cells_per_col) and curr_row - 1 < self.cells_per_col[next_col] and curr_row > 0:
                neighbours.append(self._get_ind_from_col_row(next_col, curr_row - 1))
            else:
                neighbours.append(self.num_cells + 1)

        elif curr_col < len(self.cells_per_col) / 2:
            if prev_col > -1 and curr_row - 1 < self.cells_per_col[prev_col] and curr_row > 0:
                neighbours.append(self._get_ind_from_col_row(prev_col, curr_row - 1))
            else:
                neighbours.append(self.num_cells + 1)

            if next_col < len(self.cells_per_col) and curr_row + 1 < self.cells_per_col[next_col]:
                neighbours.append(self._get_ind_from_col_row(next_col, curr_row + 1))
            else:
                neighbours.append(self.num_cells + 1)

        else:
            if prev_col > -1 and curr_row + 1 < self.cells_per_col[prev_col]:
                neighbours.append(self._get_ind_from_col_row(prev_col, curr_row + 1))
            else:
                neighbours.append(self.num_cells + 1)

            if next_col < len(self.cells_per_col) and curr_row - 1 < self.cells_per_col[next_col] and curr_row > 0:
                neighbours.append(self._get_ind_from_col_row(next_col, curr_row - 1))
            else:
                neighbours.append(self.num_cells + 1)

        return neighbours

    def reset_canvas(self):
        background(BGN_COLOUR)
        self.undo.clear()
        self.redo.clear()
        self.marbles.clear()
        self.rings.clear()
        self.num_cells = 37 #self._get_num_cells()
        self._populate_rings()

        # Setup the Marbles
        posx = CANVAS_WIDTH * 725 / 1000
        num_marbles = [10, 8, 6]
        c = ["C", "P", "Y"]
        for i, colour in enumerate(c):
            posx += self.MARBLE_RADIUS + 10
            posy = (CANVAS_HEIGHT * 180 / 900) + i * self.MARBLE_RADIUS
            for j in range(num_marbles[i]):
                marble = Marble(posx, posy, self.MARBLE_RADIUS, colour)
                self.marbles.append(marble)
                posy += self.MARBLE_RADIUS + 10

        # Setup the Rings
        self.cells_per_col = [4,5,6,7,6,5,4]
        self.cum_cells = [4,9,15,22,28,33, 37]

        for i in range(self.num_cells):
            col, row = self._get_col_row_from_ind(i)
            posx = col*(self.RING_RADIUS - 1) + CANVAS_WIDTH * 230 / 1000
            posy = (CANVAS_HEIGHT * 340 / 900) - (self.cells_per_col[col] % self.side_length) * self.RING_RADIUS / 2
            posy += row*(self.RING_RADIUS + 1)
            self.rings[i]._ring_params(x = posx,
                                       y = posy,
                                       colour=245,
                                       radius = self.RING_RADIUS)
        # self.set_removable_rings()

    def set_removable_rings(self):
        for i, ring in enumerate(self.rings):
            if ring is not None:
                neighbour_inds = self._get_neighbours(i)
                neighbour_inds.sort()
                if (self.rings[neighbour_inds[0]] is None and self.rings[neighbour_inds[1]] is None) or \
                   (self.rings[neighbour_inds[1]] is None and self.rings[neighbour_inds[2]] is None) or \
                   (self.rings[neighbour_inds[3]] is None and self.rings[neighbour_inds[5]] is None) or \
                   (self.rings[neighbour_inds[4]] is None and self.rings[neighbour_inds[5]] is None) or \
                   (self.rings[neighbour_inds[2]] is None and self.rings[neighbour_inds[2]] is None) or \
                   (self.rings[neighbour_inds[0]] is None and self.rings[neighbour_inds[1]] is None):
                    ring.is_removable = True
                else:
                    ring.is_removable = False

    def render(self):
        self.set_removable_rings()
        for ring in self.rings:
            if ring is not None:
                if ring.is_removable:
                    ring._change_colour(150)
                else:
                    ring._change_colour(245)
                ring.render()

        for marble in self.marbles:
            marble.render()


z = Zertz(4, Ring)

def setup():
    createCanvas(CANVAS_WIDTH, CANVAS_HEIGHT)
    background(BGN_COLOUR)
    z.reset_canvas()


def draw():
    fill("blue")
    background(BGN_COLOUR)
    radius = sin(frameCount / 60) * 50 + 50
    ellipse(10, 10, radius, radius)
    z.render()

    # console.log(frameRate())

def mousePressed():
    # Check if any of the marbles were clicked
    for marble in z.marbles:
        marble._clicked()

    # Check if any of the rings were clicked
    for ring in z.rings:
        if ring is not None:
            ring._clicked()

def mouseMoved():
    for i, ring in enumerate(z.rings):
        if ring is not None:
            ring._hover()
            if ring.is_hover:
                neighbour_inds = z._get_neighbours(i)
                console.log(z.rings[neighbour_inds[5]])
                for i, ring in enumerate(z.rings):
                    if i in neighbour_inds:
                        if ring is not None:
                            ring._change_colour(180)
                    else:
                        if ring is not None:
                            ring._change_colour(245)


def mouseDragged():
    # Check if any of the marbles were dragged
    for marble in z.marbles:
        marble._dragged()

def mouseReleased():
    # Check if the dragged  marble is on any ring
    for marble in z.marbles:
        if marble.is_clicked:
            for ring in z.rings:
                if ring is not None:
                    d = dist(mouseX, mouseY, ring.x, ring.y)

                    if d < ring.ring_radius/2:
                        marble.set_pos(ring.x, ring.y)
                        marble.on_ring = True
                        ring.has_marble = True
                        ring.is_removable = False
                    else:
                        marble.reset_init_pos()

    # Check if any of the rings were clicked
    for i, ring in enumerate(z.rings):
        if ring is not None:
            if ring.is_clicked and ring.is_removable:
                # console.log(i)
                z.rings[i] = None

File: JS/Zertz/zertz.js
class Zertz extends AbstractStrategyGame{
    constructor(side_length) {

        super();
        this.side_length = side_length;

        this._config_board();
        this.marbles = [];
        this.rings = [];
        this.removable_ring_ids = []; // required when listing all possible moves
        this.marble_on_board = [];

        // stacks for undo and redo
        this.undo = [];
        this.redo = [];
    }

    _config_board() {
        this.cells_per_col = [];
        this.cum_sum = [];

        for (let i = this.side_length; i < 2 * this.side_length; ++i) {
            this.cells_per_col.push(i);
        }

        for (let i = 2 * this.side_length - 2; i >= this.side_length; --i) {
            this.cells_per_col.push(i);
        }

        this.cum_sum.push(this.cells_per_col[0]);

        for (let i = 1; i < this.cells_per_col.length; ++i) {
            this.cum_sum[i] = this.cum_sum[i - 1] + this.cells_per_col[i];
        }
        this.num_cells = this.cells_per_col.reduce(function(a, b) {
            return a + b;
        }, 0);
    }

    _populate_rings() {
        for (let i = 0; i < this.num_cells; ++i) {
            this.rings.push(new Ring());
        }
        this.rings.push(null);
    }

    _get_col_row_from_ind(ind) {
        let col = 0,
            row = 0;

        for (col = 0; col < this.cum_sum.length; ++col) {
            if (ind < this.cum_sum[col]) {
                break;
            }
        }

        if (col == 0) {
            row = ind;
        } else {
            row = ind - this.cum_sum[col - 1];
        }

        return [col, row];
    }

    _get_ind_from_col_row(col, row) {
        if (col > 0) {
            return this.cum_sum[col - 1] + row;
        } else {
            return row;
        }
    }

    _get_neighbours(ind) {
        /* ........ MAIN LOGIC ..............*/
        /*
            Represent a Hexgrid as a 1D array and get the
            neighbours solely based on logic.
        */
        let neighbour_inds = [];
        let neighbours = {};

        let curr_coords = this._get_col_row_from_ind(ind);
        let cur_col = curr_coords[0],
            cur_row = curr_coords[1];

        // Top neighbour
        if (cur_row > 0) {
            neighbour_inds.push(ind - 1);
            neighbours.T = ind - 1;
        } else {
            neighbour_inds.push(this.num_cells);
            neighbours.T = this.num_cells;
        }

        // Bottom neighbour
        if (cur_row < this.cells_per_col[cur_col] - 1) {
            neighbour_inds.push(ind + 1);
            neighbours.B = ind + 1;
        } else {
            neighbour_inds.push(this.num_cells);
            neighbours.B = this.num_cells;
        }

        // Diagonal neighbours
        let prev_col = cur_col - 1;
        let next_col = cur_col + 1;

        if (prev_col > -1 && 
            cur_row < this.cells_per_col[prev_col]) {
            neighbour_inds.push(this._get_ind_from_col_row(prev_col, cur_row));

            if (cur_col <= this.cells_per_col.length / 2){
                neighbours.BL = this._get_ind_from_col_row(prev_col, cur_row);
            } else {
                neighbours.TL = this._get_ind_from_col_row(prev_col, cur_row);
            }
        } else {
            neighbour_inds.push(this.num_cells);
            if (cur_col <= this.cells_per_col.length / 2){
                neighbours.BL = this.num_cells;
            } else {
                neighbours.TL = this.num_cells;
            }
        }

        if (next_col < this.cells_per_col.length &&
            cur_row < this.cells_per_col[next_col]){
            neighbour_inds.push(this._get_ind_from_col_row(next_col, cur_row));

            if (cur_col < this.cells_per_col.length / 2){
                neighbours.TR = this._get_ind_from_col_row(prev_col, cur_row);
            } else {
                neighbours.BR = this._get_ind_from_col_row(prev_col, cur_row);
            }

        } else {
            neighbour_inds.push(this.num_cells);

            if (cur_col < this.cells_per_col.length / 2){
                neighbours.TR = this.num_cells;
            } else {
                neighbours.BR = this.num_cells;
            }
        }

        // Accounting for the center column
        if (cur_col == Math.floor(this.cells_per_col.length / 2)) {
            if (prev_col > -1 &&
                cur_row - 1 < this.cells_per_col[prev_col] &&
                cur_row > 0) {
                neighbour_inds.push(this._get_ind_from_col_row(prev_col, cur_row - 1));

                neighbours.TL = this._get_ind_from_col_row(prev_col, cur_row - 1);
            } else {
                neighbour_inds.push(this.num_cells);

                neighbours.TL = this.num_cells;
            }

            if (next_col < this.cells_per_col.length &&
                cur_row - 1 < this.cells_per_col[next_col] &&
                cur_row > 0) {
                neighbour_inds.push(this._get_ind_from_col_row(next_col, cur_row - 1));

                neighbours.TR = this._get_ind_from_col_row(next_col, cur_row - 1);
            } else {
                neighbour_inds.push(this.num_cells);

                neighbours.TR = this.num_cells;
            }
        }
        // Accounting for Left Columns
        else if (cur_col < this.cells_per_col.length / 2) {
            if (prev_col > -1 &&
                cur_row - 1 < this.cells_per_col[prev_col] &&
                cur_row > 0) {
                neighbour_inds.push(this._get_ind_from_col_row(prev_col, cur_row - 1));

                neighbours.TL = this._get_ind_from_col_row(prev_col, cur_row - 1);

            } else {
                neighbour_inds.push(this.num_cells);

                neighbours.TL = this.num_cells;
            }
            if (next_col < this.cells_per_col.length && 
                cur_row + 1 < this.cells_per_col[next_col]) {
                neighbour_inds.push(this._get_ind_from_col_row(next_col, cur_row + 1));

                neighbours.BR = this._get_ind_from_col_row(next_col, cur_row + 1);
            } else {
                neighbour_inds.push(this.num_cells);

                neighbours.BR = this.num_cells;
            }
        }
        // Accounting for the Right columns
        else {
            if (prev_col > -1 &&
                cur_row + 1 < this.cells_per_col[prev_col]) {
                neighbour_inds.push(this._get_ind_from_col_row(prev_col, cur_row + 1));

                neighbours.BL = this._get_ind_from_col_row(prev_col, cur_row + 1);
            } else {
                neighbour_inds.push(this.num_cells);

                neighbours.BL = this.num_cells;
            }

            if (next_col < this.cells_per_col.length &&
                cur_row - 1 < this.cells_per_col[next_col] &&
                cur_row > 0) {
                neighbour_inds.push(this._get_ind_from_col_row(next_col, cur_row - 1));
                
                neighbours.TR = this._get_ind_from_col_row(next_col, cur_row - 1)
            } else {
                neighbour_inds.push(this.num_cells);

                neighbours.TR = this.num_cells;
            }

        }

        return [neighbour_inds, neighbours];
    }

    _set_removable_rings() {
        this.removable_ring_ids = [];
        for (let i = 0; i < this.rings.length; ++i) {
            if (this.rings[i] != null){
                let neighbours = this._get_neighbours(i);
                neighbours = neighbours[1];

                if ((this.rings[neighbours.B] === null && this.rings[neighbours.BL] === null) ||
                    (this.rings[neighbours.B] === null && this.rings[neighbours.BR] === null) ||
                    (this.rings[neighbours.T] === null && this.rings[neighbours.TL] === null) ||
                    (this.rings[neighbours.T] === null && this.rings[neighbours.TR] === null) ||
                    (this.rings[neighbours.BL] === null && this.rings[neighbours.TL] === null) ||
                    (this.rings[neighbours.BR] === null && this.rings[neighbours.TR] === null)) {
                        if (!this.rings[i].has_marble) {
                        this.rings[i].is_removable = true;
                        this.removable_ring_ids.push(i);
                        } else{
                            this.rings[i].is_removable = false;
                        }
                    }
                    else {
                        this.rings[i].is_removable = false;
                    }

            }
        }
    }

    _get_possible_moves(){

    }

    _check_if_captured(){

    }


    _reset_canvas() {
        clear();
        background(BGN_COLOUR);
        this.rings = [];
        this.marbles = [];
        this.undo = [];
        this.redo = [];
        this._populate_rings();

        // Setup Marbles
        let posx = (CANVAS_WIDTH * 725) / 1000;
        let posy = 0;        
        let colours = [...Object.keys(MARBLE_COLOURS)];
        for (let c = 0; c < colours.length; ++c) {
            posx += MARBLE_RADIUS + 10;
            posy = (CANVAS_HEIGHT * 150) / 900 + c * MARBLE_RADIUS;
            for (let j = 0; j < NUM_MARBLES[c]; ++j) {
                this.marbles.push(new Marble(posx, posy, colours[c]));
                posy += MARBLE_RADIUS + 10;
            }
        }

        // Setup Rings
        let offset = this.side_length % 4;
        for (let i = 0; i < this.num_cells; ++i) {
            let coords = this._get_col_row_from_ind(i);
            posx = coords[0] * (RING_RADIUS - 1) + (CANVAS_WIDTH * 230) / 1000; // * (this.side_length % 4);
            posy =
                (CANVAS_HEIGHT * 340) / 900 -
                ((this.cells_per_col[coords[0]] % this.side_length) * RING_RADIUS) / 2;
            posy += coords[1] * (RING_RADIUS + 1);

            // position the hexgrid based on the sidelength
            posx -= offset * 50;
            posy -= offset * 10; 

            this.rings[i]._initialize(posx, posy);
        }

        
    }

    _render() {
        this._set_removable_rings();
        for (let ring of this.rings) {
            if (ring != null) {
                if (ring.is_removable){
                    ring.colour = [229, 219, 199];
                } else {
                    ring.colour = 245;
                }
                ring._render();
            }
        }

        for (let marble of this.marbles) {
            marble._render();
        }
    }
}